File: libs/ontology-kernel/src/domain/object_sets.rs
use std::collections::{HashSet, VecDeque};

use auth_middleware::claims::Claims;
use chrono::Utc;
use serde_json::{Map, Value, json};
use uuid::Uuid;

use crate::{
    AppState,
    domain::{
        access::{clearance_rank, marking_rank, validate_marking},
        function_runtime::{load_accessible_object_set, load_linked_objects},
    },
    models::object_set::{
        ObjectSetDefinition, ObjectSetEvaluationResponse, ObjectSetFilter, ObjectSetJoin,
        ObjectSetPolicy, ObjectSetTraversal,
    },
};

pub fn validate_object_set_definition(definition: &ObjectSetDefinition) -> Result<(), String> {
    if definition.name.trim().is_empty() {
        return Err("name is required".to_string());
    }
    for filter in &definition.filters {
        validate_filter(filter)?;
    }
    for traversal in &definition.traversals {
        validate_traversal(traversal)?;
    }
    if let Some(join) = definition.join.as_ref() {
        validate_join(join)?;
    }
    for projection in &definition.projections {
        if projection.trim().is_empty() {
            return Err("projections cannot contain empty values".to_string());
        }
    }
    validate_policy(&definition.policy)?;
    Ok(())
}

pub async fn evaluate_object_set(
    state: &AppState,
    claims: &Claims,
    definition: &ObjectSetDefinition,
    limit: usize,
    materialized: bool,
) -> Result<ObjectSetEvaluationResponse, String> {
    validate_object_set_definition(definition)?;
    enforce_object_set_policy(claims, &definition.policy)?;

    let base_objects =
        load_accessible_object_set(state, claims, definition.base_object_type_id).await?;
    let filtered = base_objects
        .into_iter()
        .filter(|object| allows_marking(claims, &definition.policy, object))
        .filter(|object| matches_filters(object, &definition.filters))
        .collect::<Vec<_>>();

    let mut traversal_neighbor_count = 0usize;
    let secondary_rows = if let Some(join) = definition.join.as_ref() {
        Some(load_accessible_object_set(state, claims, join.secondary_object_type_id).await?)
    } else {
        None
    };

    let mut rows = Vec::new();
    for base in &filtered {
        let neighbors = resolve_traversals(state, claims, base, &definition.traversals).await?;
        traversal_neighbor_count += neighbors.len();
        let seed_row = json!({
            "base": base,
            "neighbors": neighbors,
            "what_if_label": definition.what_if_label,
        });

        if let Some(join) = definition.join.as_ref() {
            let joined = secondary_rows
                .as_ref()
                .into_iter()
                .flat_map(|rows| rows.iter())
                .filter(|candidate| join_matches(base, candidate, join))
                .cloned()
                .collect::<Vec<_>>();

            if joined.is_empty() {
                if join.join_kind.eq_ignore_ascii_case("left") {
                    rows.push(augment_row_with_join(seed_row.clone(), Value::Null));
                }
                continue;
            }

            for item in joined {
                rows.push(augment_row_with_join(seed_row.clone(), item));
            }
        } else {
            rows.push(seed_row);
        }
    }

    let total_rows = rows.len();
    let limited_rows = rows
        .into_iter()
        .take(limit)
        .map(|row| project_row(row, &definition.projections))
        .collect::<Vec<_>>();

    Ok(ObjectSetEvaluationResponse {
        object_set: definition.clone(),
        total_base_matches: filtered.len(),
        total_rows,
        traversal_neighbor_count,
        rows: limited_rows,
        generated_at: Utc::now(),
        materialized,
    })
}

fn validate_filter(filter: &ObjectSetFilter) -> Result<(), String> {
    if filter.field.trim().is_empty() {
        return Err("filters require a field".to_string());
    }
    match filter.operator.as_str() {
        "equals" | "not_equals" | "contains" | "in" | "exists" | "gte" | "lte" => Ok(()),
        other => Err(format!("unsupported filter operator '{other}'")),
    }
}

fn validate_traversal(traversal: &ObjectSetTraversal) -> Result<(), String> {
    match traversal.direction.as_str() {
        "outbound" | "inbound" | "both" => {}
        other => return Err(format!("unsupported traversal direction '{other}'")),
    }
    if traversal.max_hops <= 0 || traversal.max_hops > 4 {
        return Err("traversal.max_hops must be between 1 and 4".to_string());
    }
    Ok(())
}

fn validate_join(join: &ObjectSetJoin) -> Result<(), String> {
    if join.left_field.trim().is_empty() || join.right_field.trim().is_empty() {
        return Err("join fields cannot be empty".to_string());
    }
    match join.join_kind.as_str() {
        "inner" | "left" => Ok(()),
        other => Err(format!("unsupported join kind '{other}'")),
    }
}

fn validate_policy(policy: &ObjectSetPolicy) -> Result<(), String> {
    for marking in &policy.allowed_markings {
        validate_marking(marking)?;
    }
    if let Some(minimum_clearance) = policy.minimum_clearance.as_deref() {
        validate_marking(minimum_clearance)?;
    }
    Ok(())
}

fn enforce_object_set_policy(claims: &Claims, policy: &ObjectSetPolicy) -> Result<(), String> {
    if policy.deny_guest_sessions && claims.is_guest_session() {
        return Err("forbidden: object set blocks guest sessions".to_string());
    }
    if let Some(minimum_clearance) = policy.minimum_clearance.as_deref() {
        let required = marking_rank(minimum_clearance)
            .ok_or_else(|| format!("invalid minimum clearance '{minimum_clearance}'"))?;
        if clearance_rank(claims) < required {
            return Err(
                "forbidden: insufficient classification clearance for object set".to_string(),
            );
        }
    }
    if let Some(restricted_view_id) = policy.required_restricted_view_id {
        let allowed = claims.restricted_view_ids();
        if !allowed.contains(&restricted_view_id) && !claims.has_role("admin") {
            return Err(
                "forbidden: required restricted view is not present in the session".to_string(),
            );
        }
    }
    Ok(())
}

fn allows_marking(claims: &Claims, policy: &ObjectSetPolicy, object: &Value) -> bool {
    let Some(marking) = resolve_path(object, "marking").and_then(Value::as_str) else {
        return false;
    };
    if !claims.allows_marking(marking) {
        return false;
    }
    if policy.allowed_markings.is_empty() {
        return true;
    }
    policy
        .allowed_markings
        .iter()
        .any(|candidate| candidate.eq_ignore_ascii_case(marking))
}

fn matches_filters(object: &Value, filters: &[ObjectSetFilter]) -> bool {
    filters.iter().all(|filter| {
        let actual = resolve_path(object, &filter.field);
        match filter.operator.as_str() {
            "equals" => actual == Some(&filter.value),
            "not_equals" => actual != Some(&filter.value),
            "contains" => contains_value(actual, &filter.value),
            "in" => filter
                .value
                .as_array()
                .map(|values| {
                    actual.is_some_and(|current| values.iter().any(|item| item == current))
                })
                .unwrap_or(false),
            "exists" => actual.is_some() == filter.value.as_bool().unwrap_or(true),
            "gte" => compare_values(actual, Some(&filter.value)).is_some_and(|value| value >= 0),
            "lte" => compare_values(actual, Some(&filter.value)).is_some_and(|value| value <= 0),
            _ => false,
        }
    })
}

async fn resolve_traversals(
    state: &AppState,
    claims: &Claims,
    base: &Value,
    traversals: &[ObjectSetTraversal],
) -> Result<Vec<Value>, String> {
    if traversals.is_empty() {
        return Ok(Vec::new());
    }

    let Some(base_id) = resolve_path(base, "id")
        .and_then(Value::as_str)
        .and_then(|value| Uuid::parse_str(value).ok())
    else {
        return Ok(Vec::new());
    };

    let mut all_neighbors = Vec::new();
    for traversal in traversals {
        let mut seen = HashSet::new();
        let mut queue = VecDeque::from([(base_id, 0i32)]);
        seen.insert(base_id);
        while let Some((current_id, depth)) = queue.pop_front() {
            if depth >= traversal.max_hops {
                continue;
            }
            let linked = load_linked_objects(state, claims, current_id).await?;
            for link in linked {
                if !traversal_matches(&link, traversal) {
                    continue;
                }
                let Some(object_id) = resolve_path(&link, "object.id")
                    .and_then(Value::as_str)
                    .and_then(|value| Uuid::parse_str(value).ok())
                else {
                    continue;
                };
                if !seen.insert(object_id) {
                    continue;
                }
                queue.push_back((object_id, depth + 1));
                all_neighbors.push(link);
            }
        }
    }

    Ok(all_neighbors)
}

fn traversal_matches(link: &Value, traversal: &ObjectSetTraversal) -> bool {
    if traversal.direction != "both" {
        let direction = resolve_path(link, "direction").and_then(Value::as_str);
        if direction != Some(traversal.direction.as_str()) {
            return false;
        }
    }
    if let Some(link_type_id) = traversal.link_type_id {
        let actual = resolve_path(link, "link_type_id")
            .and_then(Value::as_str)
            .and_then(|value| Uuid::parse_str(value).ok());
        if actual != Some(link_type_id) {
            return false;
        }
    }
    if let Some(target_object_type_id) = traversal.target_object_type_id {
        let actual = resolve_path(link, "object.object_type_id")
            .and_then(Value::as_str)
            .and_then(|value| Uuid::parse_str(value).ok());
        if actual != Some(target_object_type_id) {
            return false;
        }
    }
    true
}

fn join_matches(base: &Value, candidate: &Value, join: &ObjectSetJoin) -> bool {
    let left = resolve_path(base, &join.left_field);
    let right = resolve_path(candidate, &join.right_field);
    left.is_some() && left == right
}

fn augment_row_with_join(row: Value, joined: Value) -> Value {
    let mut object = row.as_object().cloned().unwrap_or_default();
    object.insert("joined".to_string(), joined);
    Value::Object(object)
}

fn project_row(row: Value, projections: &[String]) -> Value {
    if projections.is_empty() {
        return row;
    }

    let mut projected = Map::new();
    for projection in projections {
        projected.insert(
            projection.clone(),
            resolve_projection_value(&row, projection)
                .cloned()
                .unwrap_or(Value::Null),
        );
    }
    Value::Object(projected)
}

fn resolve_projection_value<'a>(row: &'a Value, projection: &str) -> Option<&'a Value> {
    if projection.starts_with("base.")
        || projection.starts_with("joined.")
        || projection.starts_with("neighbors.")
    {
        return resolve_path(row, projection);
    }

    resolve_path(row, projection).or_else(|| resolve_path(row, &format!("base.{projection}")))
}

fn resolve_path<'a>(value: &'a Value, path: &str) -> Option<&'a Value> {
    if path.trim().is_empty() {
        return Some(value);
    }

    if !path.contains('.') {
        return value
            .get(path)
            .or_else(|| {
                value
                    .get("properties")
                    .and_then(|properties| properties.get(path))
            })
            .or_else(|| value.get("base").and_then(|base| resolve_path(base, path)));
    }

    let mut current = value;
    for segment in path.split('.') {
        current = current.get(segment)?;
    }
    Some(current)
}

fn contains_value(actual: Option<&Value>, expected: &Value) -> bool {
    match (actual, expected) {
        (Some(Value::String(current)), Value::String(candidate)) => current.contains(candidate),
        (Some(Value::Array(items)), candidate) => items.iter().any(|item| item == candidate),
        _ => false,
    }
}

fn compare_values(left: Option<&Value>, right: Option<&Value>) -> Option<i8> {
    let left = left?;
    let right = right?;
    if let (Some(l), Some(r)) = (left.as_f64(), right.as_f64()) {
        return Some(match l.partial_cmp(&r)? {
            std::cmp::Ordering::Less => -1,
            std::cmp::Ordering::Equal => 0,
            std::cmp::Ordering::Greater => 1,
        });
    }
    if let (Some(l), Some(r)) = (left.as_str(), right.as_str()) {
        return Some(match l.cmp(r) {
            std::cmp::Ordering::Less => -1,
            std::cmp::Ordering::Equal => 0,
            std::cmp::Ordering::Greater => 1,
        });
    }
    None
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{compare_values, matches_filters, project_row, resolve_path};
    use crate::models::object_set::ObjectSetFilter;

    #[test]
    fn object_set_filters_support_properties_and_marking() {
        let object = json!({
            "id": "00000000-0000-0000-0000-000000000001",
            "marking": "confidential",
            "properties": {
                "status": "active",
                "score": 99
            }
        });

        assert!(matches_filters(
            &object,
            &[
                ObjectSetFilter {
                    field: "status".to_string(),
                    operator: "equals".to_string(),
                    value: json!("active"),
                },
                ObjectSetFilter {
                    field: "marking".to_string(),
                    operator: "equals".to_string(),
                    value: json!("confidential"),
                },
                ObjectSetFilter {
                    field: "score".to_string(),
                    operator: "gte".to_string(),
                    value: json!(70),
                },
            ],
        ));
    }

    #[test]
    fn projections_prefer_wrapper_and_base_paths() {
        let row = json!({
            "base": {
                "id": "1",
                "properties": {
                    "name": "Case-7"
                }
            },
            "joined": {
                "properties": {
                    "owner": "Alice"
                }
            }
        });

        let projected = project_row(
            row,
            &[
                "id".to_string(),
                "base.properties.name".to_string(),
                "joined.properties.owner".to_string(),
            ],
        );

        assert_eq!(projected["id"], json!("1"));
        assert_eq!(projected["base.properties.name"], json!("Case-7"));
        assert_eq!(projected["joined.properties.owner"], json!("Alice"));
    }

    #[test]
    fn compare_values_handles_numbers_and_strings() {
        assert_eq!(compare_values(Some(&json!(3)), Some(&json!(7))), Some(-1));
        assert_eq!(
            compare_values(Some(&json!("b")), Some(&json!("a"))),
            Some(1)
        );
        assert_eq!(
            resolve_path(&json!({"properties": {"kind": "case"}}), "kind"),
            Some(&json!("case"))
        );
    }
}

File: libs/ontology-kernel/domain/object_sets.go
// Object set definition validation + filter / projection helpers.
//
// Mirrors `libs/ontology-kernel/src/domain/object_sets.rs` for the
// subset that does NOT depend on `domain::function_runtime`. The two
// store-bound helpers (`evaluate_object_set` and `resolve_traversals`,
// which call `load_accessible_object_set` / `load_linked_objects`)
// will land in iter 7c₅ alongside the function_runtime port.

package domain

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"

	authmw "github.com/openfoundry/openfoundry-go/libs/auth-middleware"
	ontologykernel "github.com/openfoundry/openfoundry-go/libs/ontology-kernel"
	"github.com/openfoundry/openfoundry-go/libs/ontology-kernel/models"
)

// EvaluateObjectSet mirrors `pub async fn evaluate_object_set`. It validates
// the definition and policy, loads tenant-scoped accessible objects through the
// function runtime helpers, applies filters/traversals/joins/projections, and
// returns the same response envelope as Rust.
func EvaluateObjectSet(ctx context.Context, state *ontologykernel.AppState, claims *authmw.Claims, definition models.ObjectSetDefinition, limit int, materialized bool) (models.ObjectSetEvaluationResponse, error) {
	if err := ValidateObjectSetDefinition(definition); err != nil {
		return models.ObjectSetEvaluationResponse{}, err
	}
	if err := EnforceObjectSetPolicy(claims, definition.Policy); err != nil {
		return models.ObjectSetEvaluationResponse{}, err
	}
	if limit < 0 {
		limit = 0
	}

	baseObjects, err := LoadAccessibleObjectSet(ctx, state, claims, definition.BaseObjectTypeID)
	if err != nil {
		return models.ObjectSetEvaluationResponse{}, err
	}
	filtered := make([]json.RawMessage, 0, len(baseObjects))
	for _, object := range baseObjects {
		if AllowsObjectSetMarking(claims, definition.Policy, object) && MatchesObjectSetFilters(object, definition.Filters) {
			filtered = append(filtered, object)
		}
	}

	var secondaryRows []json.RawMessage
	if definition.Join != nil {
		secondaryRows, err = LoadAccessibleObjectSet(ctx, state, claims, definition.Join.SecondaryObjectTypeID)
		if err != nil {
			return models.ObjectSetEvaluationResponse{}, err
		}
	}

	rows := []json.RawMessage{}
	traversalNeighborCount := 0
	for _, base := range filtered {
		neighbors, err := resolveObjectSetTraversals(ctx, state, claims, base, definition.Traversals)
		if err != nil {
			return models.ObjectSetEvaluationResponse{}, err
		}
		traversalNeighborCount += len(neighbors)
		seed, _ := json.Marshal(map[string]any{
			"base":          json.RawMessage(base),
			"neighbors":     neighbors,
			"what_if_label": definition.WhatIfLabel,
		})

		if definition.Join == nil {
			rows = append(rows, seed)
			continue
		}

		joined := []json.RawMessage{}
		for _, candidate := range secondaryRows {
			if JoinMatches(base, candidate, *definition.Join) {
				joined = append(joined, candidate)
			}
		}
		if len(joined) == 0 {
			if strings.EqualFold(definition.Join.JoinKind, "left") {
				rows = append(rows, AugmentObjectSetRowWithJoin(seed, json.RawMessage("null")))
			}
			continue
		}
		for _, item := range joined {
			rows = append(rows, AugmentObjectSetRowWithJoin(seed, item))
		}
	}

	totalRows := len(rows)
	if limit > len(rows) {
		limit = len(rows)
	}
	limited := make([]json.RawMessage, 0, limit)
	for _, row := range rows[:limit] {
		limited = append(limited, ProjectObjectSetRow(row, definition.Projections))
	}

	return models.ObjectSetEvaluationResponse{
		ObjectSet:              definition,
		TotalBaseMatches:       len(filtered),
		TotalRows:              totalRows,
		TraversalNeighborCount: traversalNeighborCount,
		Rows:                   limited,
		GeneratedAt:            time.Now().UTC(),
		Materialized:           materialized,
	}, nil
}

func resolveObjectSetTraversals(ctx context.Context, state *ontologykernel.AppState, claims *authmw.Claims, base json.RawMessage, traversals []models.ObjectSetTraversal) ([]json.RawMessage, error) {
	if len(traversals) == 0 {
		return []json.RawMessage{}, nil
	}
	idRaw := ResolveObjectSetPath(base, "id")
	idText, ok := jsonString(idRaw)
	if !ok {
		return []json.RawMessage{}, nil
	}
	id, err := uuid.Parse(idText)
	if err != nil {
		return []json.RawMessage{}, nil
	}
	frontier := []uuid.UUID{id}
	seen := map[uuid.UUID]bool{id: true}
	resolved := []json.RawMessage{}
	maxHops := int32(0)
	for _, traversal := range traversals {
		if traversal.MaxHops > maxHops {
			maxHops = traversal.MaxHops
		}
	}
	for hop := int32(0); hop < maxHops && len(frontier) > 0; hop++ {
		next := []uuid.UUID{}
		for _, current := range frontier {
			linked, err := LoadLinkedObjects(ctx, state, claims, current)
			if err != nil {
				return nil, err
			}
			for _, link := range linked {
				for _, traversal := range traversals {
					if traversal.MaxHops <= hop || !TraversalMatches(link, traversal) {
						continue
					}
					resolved = append(resolved, link)
					neighborText, ok := jsonString(ResolveObjectSetPath(link, "object.id"))
					if !ok {
						continue
					}
					neighborID, err := uuid.Parse(neighborText)
					if err == nil && !seen[neighborID] {
						seen[neighborID] = true
						next = append(next, neighborID)
					}
				}
			}
		}
		frontier = next
	}
	return resolved, nil
}

// ValidateObjectSetDefinition mirrors `pub fn validate_object_set_definition`.
// Returns nil on success or the verbatim Rust error string on the
// first failing rule.
func ValidateObjectSetDefinition(definition models.ObjectSetDefinition) error {
	if strings.TrimSpace(definition.Name) == "" {
		return fmt.Errorf("name is required")
	}
	for _, filter := range definition.Filters {
		if err := ValidateObjectSetFilter(filter); err != nil {
			return err
		}
	}
	for _, traversal := range definition.Traversals {
		if err := ValidateObjectSetTraversal(traversal); err != nil {
			return err
		}
	}
	if definition.Join != nil {
		if err := ValidateObjectSetJoin(*definition.Join); err != nil {
			return err
		}
	}
	for _, projection := range definition.Projections {
		if strings.TrimSpace(projection) == "" {
			return fmt.Errorf("projections cannot contain empty values")
		}
	}
	return ValidateObjectSetPolicy(definition.Policy)
}

// ValidateObjectSetFilter mirrors `fn validate_filter`.
func ValidateObjectSetFilter(filter models.ObjectSetFilter) error {
	if strings.TrimSpace(filter.Field) == "" {
		return fmt.Errorf("filters require a field")
	}
	switch filter.Operator {
	case "equals", "not_equals", "contains", "in", "exists", "gte", "lte":
		return nil
	default:
		return fmt.Errorf("unsupported filter operator '%s'", filter.Operator)
	}
}

// ValidateObjectSetTraversal mirrors `fn validate_traversal`.
func ValidateObjectSetTraversal(traversal models.ObjectSetTraversal) error {
	switch traversal.Direction {
	case "outbound", "inbound", "both":
	default:
		return fmt.Errorf("unsupported traversal direction '%s'", traversal.Direction)
	}
	if traversal.MaxHops <= 0 || traversal.MaxHops > 4 {
		return fmt.Errorf("traversal.max_hops must be between 1 and 4")
	}
	return nil
}

// ValidateObjectSetJoin mirrors `fn validate_join`.
func ValidateObjectSetJoin(join models.ObjectSetJoin) error {
	if strings.TrimSpace(join.LeftField) == "" || strings.TrimSpace(join.RightField) == "" {
		return fmt.Errorf("join fields cannot be empty")
	}
	switch join.JoinKind {
	case "inner", "left":
		return nil
	default:
		return fmt.Errorf("unsupported join kind '%s'", join.JoinKind)
	}
}

// ValidateObjectSetPolicy mirrors `fn validate_policy`. Calls
// [ValidateMarking] (access.go) for both the allowlist entries and
// the optional minimum_clearance.
func ValidateObjectSetPolicy(policy models.ObjectSetPolicy) error {
	for _, marking := range policy.AllowedMarkings {
		if err := ValidateMarking(marking); err != nil {
			return err
		}
	}
	if policy.MinimumClearance != nil {
		if err := ValidateMarking(*policy.MinimumClearance); err != nil {
			return err
		}
	}
	return nil
}

// EnforceObjectSetPolicy mirrors `fn enforce_object_set_policy`.
// Pure-claims enforcement; no IO.
func EnforceObjectSetPolicy(claims *authmw.Claims, policy models.ObjectSetPolicy) error {
	if policy.DenyGuestSessions && claims != nil && claims.IsGuestSession() {
		return fmt.Errorf("forbidden: object set blocks guest sessions")
	}
	if policy.MinimumClearance != nil {
		required, ok := MarkingRank(*policy.MinimumClearance)
		if !ok {
			return fmt.Errorf("invalid minimum clearance '%s'", *policy.MinimumClearance)
		}
		if ClearanceRank(claims) < required {
			return fmt.Errorf("forbidden: insufficient classification clearance for object set")
		}
	}
	if policy.RequiredRestrictedViewID != nil && claims != nil {
		if !claims.HasRole("admin") {
			allowed := claims.RestrictedViewIDs()
			matched := false
			for _, id := range allowed {
				if id == *policy.RequiredRestrictedViewID {
					matched = true
					break
				}
			}
			if !matched {
				return fmt.Errorf("forbidden: required restricted view is not present in the session")
			}
		}
	}
	return nil
}

// AllowsObjectSetMarking mirrors `fn allows_marking`. Receives the
// JSON-encoded object (Rust `&Value`) and inspects the `marking`
// field. Empty marking → reject; otherwise both the user's allowlist
// and the policy's allowlist must accept it.
func AllowsObjectSetMarking(claims *authmw.Claims, policy models.ObjectSetPolicy, object json.RawMessage) bool {
	marking, ok := jsonString(ResolveObjectSetPath(object, "marking"))
	if !ok || marking == "" {
		return false
	}
	if claims == nil || !claims.AllowsMarking(marking) {
		return false
	}
	if len(policy.AllowedMarkings) == 0 {
		return true
	}
	for _, candidate := range policy.AllowedMarkings {
		if strings.EqualFold(candidate, marking) {
			return true
		}
	}
	return false
}

// MatchesObjectSetFilters mirrors `fn matches_filters`. Walks every
// filter and returns true when ALL pass.
func MatchesObjectSetFilters(object json.RawMessage, filters []models.ObjectSetFilter) bool {
	for _, f := range filters {
		actual := ResolveObjectSetPath(object, f.Field)
		switch f.Operator {
		case "equals":
			if actual == nil || !objectSetRawEqual(actual, f.Value) {
				return false
			}
		case "not_equals":
			if actual != nil && objectSetRawEqual(actual, f.Value) {
				return false
			}
			if actual == nil {
				// Rust: actual != Some(filter.value) is true when
				// actual is None — so the predicate passes.
			}
		case "contains":
			if !rawContainsValue(actual, f.Value) {
				return false
			}
		case "in":
			items, ok := jsonArray(f.Value)
			if !ok {
				return false
			}
			if actual == nil {
				return false
			}
			match := false
			for _, item := range items {
				if objectSetRawEqual(item, actual) {
					match = true
					break
				}
			}
			if !match {
				return false
			}
		case "exists":
			expected := true
			var b bool
			if err := json.Unmarshal(f.Value, &b); err == nil {
				expected = b
			}
			if (actual != nil) != expected {
				return false
			}
		case "gte":
			c, ok := compareJSON(actual, f.Value)
			if !ok || c < 0 {
				return false
			}
		case "lte":
			c, ok := compareJSON(actual, f.Value)
			if !ok || c > 0 {
				return false
			}
		default:
			return false
		}
	}
	return true
}

// TraversalMatches mirrors `fn traversal_matches`.
func TraversalMatches(link json.RawMessage, traversal models.ObjectSetTraversal) bool {
	if traversal.Direction != "both" {
		dir, ok := jsonString(ResolveObjectSetPath(link, "direction"))
		if !ok || dir != traversal.Direction {
			return false
		}
	}
	if traversal.LinkTypeID != nil {
		s, ok := jsonString(ResolveObjectSetPath(link, "link_type_id"))
		if !ok {
			return false
		}
		if s != traversal.LinkTypeID.String() {
			return false
		}
	}
	if traversal.TargetObjectTypeID != nil {
		s, ok := jsonString(ResolveObjectSetPath(link, "object.object_type_id"))
		if !ok {
			return false
		}
		if s != traversal.TargetObjectTypeID.String() {
			return false
		}
	}
	return true
}

// JoinMatches mirrors `fn join_matches`. Both sides must resolve and
// compare equal.
func JoinMatches(base, candidate json.RawMessage, join models.ObjectSetJoin) bool {
	left := ResolveObjectSetPath(base, join.LeftField)
	right := ResolveObjectSetPath(candidate, join.RightField)
	return left != nil && objectSetRawEqual(left, right)
}

// AugmentObjectSetRowWithJoin mirrors `fn augment_row_with_join`.
// Re-marshals the row map with the `joined` field set.
func AugmentObjectSetRowWithJoin(row, joined json.RawMessage) json.RawMessage {
	obj, ok := jsonObject(row)
	if !ok {
		obj = map[string]json.RawMessage{}
	}
	obj["joined"] = joined
	out, _ := json.Marshal(obj)
	return out
}

// ProjectObjectSetRow mirrors `fn project_row`. Empty projections
// passes the row through unchanged.
func ProjectObjectSetRow(row json.RawMessage, projections []string) json.RawMessage {
	if len(projections) == 0 {
		return row
	}
	out := map[string]json.RawMessage{}
	for _, projection := range projections {
		v := resolveProjectionValue(row, projection)
		if v == nil {
			out[projection] = json.RawMessage("null")
			continue
		}
		out[projection] = v
	}
	res, _ := json.Marshal(out)
	return res
}

// resolveProjectionValue mirrors `fn resolve_projection_value`. Tries
// the wrapper-rooted path first; if absent and the projection is not
// already nested under base/joined/neighbors, retries under
// `base.<projection>`.
func resolveProjectionValue(row json.RawMessage, projection string) json.RawMessage {
	if strings.HasPrefix(projection, "base.") ||
		strings.HasPrefix(projection, "joined.") ||
		strings.HasPrefix(projection, "neighbors.") {
		return ResolveObjectSetPath(row, projection)
	}
	if v := ResolveObjectSetPath(row, projection); v != nil {
		return v
	}
	return ResolveObjectSetPath(row, "base."+projection)
}

// ResolveObjectSetPath mirrors `fn resolve_path`. Empty path returns
// the value itself; single-segment paths fall back to
// `properties.<seg>` and then `base.<seg>`; dotted paths walk each
// segment in order.
func ResolveObjectSetPath(value json.RawMessage, path string) json.RawMessage {
	if strings.TrimSpace(path) == "" {
		return value
	}
	if !strings.Contains(path, ".") {
		obj, ok := jsonObject(value)
		if !ok {
			return nil
		}
		if v, ok := obj[path]; ok {
			return v
		}
		if props, ok := obj["properties"]; ok {
			pp, ok := jsonObject(props)
			if ok {
				if v, ok := pp[path]; ok {
					return v
				}
			}
		}
		if base, ok := obj["base"]; ok {
			return ResolveObjectSetPath(base, path)
		}
		return nil
	}
	current := value
	for _, seg := range strings.Split(path, ".") {
		obj, ok := jsonObject(current)
		if !ok {
			return nil
		}
		next, ok := obj[seg]
		if !ok {
			return nil
		}
		current = next
	}
	return current
}

// objectSetRawEqual mirrors `Value == Value` after both sides have been
// canonicalised through Go's JSON decode/encode pair (so whitespace
// differences don't matter).
func objectSetRawEqual(a, b json.RawMessage) bool {
	return jsonEqual(a, b)
}

// rawContainsValue mirrors `fn contains_value`. String contains
// substring; array contains element.
func rawContainsValue(actual, expected json.RawMessage) bool {
	if actual == nil {
		return false
	}
	if as, ok := jsonString(actual); ok {
		es, ok := jsonString(expected)
		if !ok {
			return false
		}
		return strings.Contains(as, es)
	}
	if items, ok := jsonArray(actual); ok {
		for _, item := range items {
			if objectSetRawEqual(item, expected) {
				return true
			}
		}
		return false
	}
	return false
}

// (Avoid duplicate utility names with `domain.access.go` /
// `submission_eval.go`. The local helpers above are written in terms
// of the shared [jsonString], [jsonArray], [jsonObject], [jsonEqual],
// [compareJSON] implementations declared in submission_eval.go.)

// ensureBytesContainsAccessible — sanity import to avoid an unused
// compile error if a future trim drops the bytes import. The current
// implementations don't need it so the import was removed; this stub
// is a safety net for refactors.
var _ = bytes.Contains
